services/src/main/java/org/keycloak/services/DefaultKeycloakSession.java

```diff
@@ -452,28 +454,68 @@ public ClientPolicyManager clientPolicy() {
         return clientPolicyManager;
     }
 
+    private static final Logger LOG = Logger.getLogger(DefaultKeycloakSession.class);
+
     @Override
     public void close() {
+        if (LOG.isDebugEnabled()) {
+            LOG.debugf("Closing %s%s%s", this,
+              getTransactionManager().isActive() ? " (transaction active" + (getTransactionManager().getRollbackOnly() ? ", ROLLBACK-ONLY" : "") + ")" : "",
+              StackUtil.getShortStackTrace());
+        }
+
         if (closed) {
-            throw new IllegalStateException("Illegal call to #close() on already closed KeycloakSession");
+            throw new IllegalStateException("Illegal call to #close() on already closed " + this);
+        }
+
+        RuntimeException re = closeTransactionManager();
```

Comment:
Did we use to commit the transaction prior to closing the session (and providers) or the other way around?

Reply:
First the transaction manager was closed, then session was closed, see https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/services/filters/AbstractRequestFilter.java#L58-L69